### ISSUE TITLE
[Spike] Save as PDF option

### ISSUE THREAD

**Issue by sebast889** [mozilla-mobile/fenix#3709]
### Why/User Benefit/User Problem

To easily save a web page in PDF for formatting, portability and saving a local copy for later reading.

### What/Requirements

Save as PDF option to save the current webpage as a PDF file

### Acceptance Criteria (how do I know when I’m done?)

Feature parity with Fennec to save the current webpage as a PDF file

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-12323)

**Comment by yoasif** [comment]
49 thumbs up for this issue on [Google Play](https://play.google.com/store/apps/details?id=org.mozilla.firefox_beta&hl=en_US&reviewId=gp%3AAOqpTOGTU2MGPN8b935kMZSv16GMppQPFNKzCN3WYmzpFY5aj9-egUDPgZ6jyVYJt2v354Fi0Ln9S_0mo5ldYQ).

>A lot of tools I use disappeared with the changes made. Especially, print and Save as PDF. I do a lot of research and want to save items as documents for writing up research results, etc. The changes you made don't save the document as anything!

**Comment by vesta0** [comment]
Let's do an `s` investigation to determine dependencies.

**Comment by Poopooracoocoo** [comment]
There are extensions like GoFullPage that let you save a page as an image. Safari lets you save a page as a .webarchive. Desktop browsers let you download their html bundles or let you save an image through the developer tools.

**Comment by cadeyrn** [comment]
> Desktop browsers let you download their html bundles or let you save an image through the developer tools.

Firefox 81 for desktop let you save webpages as PDF. 😉

**Comment by Poopooracoocoo** [comment]
but you could always print to pdf :P

**Comment by cadeyrn** [comment]
> but you could always print to pdf :P

Not within Firefox, it's a new feature, see also https://mozilla.invisionapp.com/share/MFWMR6H5X3K#/screens/411381710. Except you use macOS because Firefox used to use the native macOS printing dialog and print to PDF is also a macOS feature.

**Comment by benklaasen** [comment]
> There are extensions like GoFullPage that let you save a page as an image. Safari lets you save a page as a .webarchive. Desktop browsers let you download their html bundles or let you save an image through the developer tools.

None of these options currently work on Firefox for Android.

**Comment by Poopooracoocoo** [comment]
I was thinking of different formats to PDFs that Mozilla could also implement.

**Comment by FunkyBrewster** [comment]
As of 8/28/2020, FF on Android has lost the ability to save pages to PDF, which was really useful. It's a shame that v79 was released in such a premature state. This is just one example of that. Also, to the people commenting about Print to PDF or using addons, that is irrelevant. There is no Print to PDF on mobile, and all but 9 addons are disabled in v79.

**Comment by TexSon** [comment]
FunkyBrewster is correct. Many of us depend on being able to Save to PDF for many reasons. From this history, that has obviously been understood for quite some time. Failure to include this critical feature is unforgivable.

**Comment by headbank** [comment]
I don't know whether this is Android or vendor-specific functionality, but on my Android 10 Samsung phone the built-in print service has a Print to PDF option. If #10217 progresses, this option will be exposed without Fenix having to do anything PDF-specific I guess.

Chrome exposes the OS-level print facility through the Share menu, but Fenix currently does not (seems like this would be the easiest route to enable the functionality).

I should mention that other Android apps may appear here that can handle saving the web page to PDF: Google Drive is there for one, and a PDF viewer I use (Xodo) is also there (however the PDF shared from Fenix in this way is very poorly scaled so not a great solution). Just if you're desperate for PDF printing this option might be worth a look as a stopgap.

**Comment by benklaasen** [comment]
You can install Fennec from F-Droid, at least until Fennec is updated to Fenix. Hopefully that won't happen for a while yet. https://f-droid.org/en/packages/org.mozilla.fennec_fdroid/

Your add-ons will continue to function and "Save as PDF" works. Firefox Sync works in Fennec, but you'll most likely need to export your add-on config customisations, if they have such functionality.

**Comment by FunkyBrewster** [comment]
@headbank 
> I don't know whether this is Android or vendor-specific functionality, but on my Android 10 Samsung phone the built-in print service has a Print to PDF option. 

I found the Print Service on my phone, but it doesn't have Print to PDF.

> I should mention that other Android apps may appear in that Share menu that can handle saving the web page to PDF: Google Drive is there for one, and a PDF viewer I use (Xodo) is also there 

Thanks for the information. I didn't know apps could use the Share menu that way so I'll look into that for now.  

(OT: There are several apps already offering to do that on my share menu but when you click them they're adds for subscription services. How annoying.)

**Comment by FunkyBrewster** [comment]
@cadeyrn 
> Firefox 81 for desktop let you save webpages as PDF. 😉

Do you even understand what we are talking about? You keep thumb-downing posts describing why this is an issue and posting off-topic comments about desktop OSes. That's not helping.

**Comment by cadeyrn** [comment]
> Do you even understand what we are talking about?

Yes, I know this very well. But obviously you did not understand my comment. It was an **answer to the comment from another user** above this comment. 😒

**Comment by madb1lly** [comment]
Hi all,

Let's get back to the topic and not discussing each other personally 🙂

Suggesting workarounds is helpful in the short term but doesn't address the problem that from the perspective of most users this was a feature previously available in Firefox for Android which has now been removed and many (this issue seems to have a lot of support) want to have that feature back, otherwise Firefox for Android will not remain as useful for them.

Cheers 🙂

**Comment by headbank** [comment]
My mention of a present workaround was really incidental (and not effective in the instance I looked at, as I stated).

My point was that Android (and on further digging it _is_ stock Android [according to this](https://www.techotopia.com/index.php/Printing_with_the_Android_6_Printing_Framework#Save_as_PDF), from at least Android 6 - so @FunkyBrewster I'm not sure why you're not seeing it unless your Android is earlier or your OEM has disabled it) already has PDF-print capability. Therefore it points to a very cheap route to "restoring" the functionality that, unlike most users, you'll understand _Fenix_ has never had.

**Comment by madb1lly** [comment]
Hi @headbank,

Thanks for the explanation, I understand you better now, please excuse me. On my Nokia 8, android 9, supposedly "stock" I can't print direct to PDF as far as I can tell. I do have the print service, but Print to PDF is not shown as one of the available "printers" (but maybe it shouldn't be either, I don't know).

That techtopia page is a bit unclear as to whether the print to PDF service is built into AOSP or if it is optional. I did find this, which might be what is referred to: https://developer.android.com/reference/android/print/pdf/package-summary

When I open a file in Google Drive app and select print I can save as PDF, but I think this might be an app-specific feature. When I ask to see all printers in that app it doesn't show me print to PDF.

Cheers 🙂

**Comment by FunkyBrewster** [comment]
So to summarize, you can work around this issue by using the Share function to another app that will output to pdf. If you're lucky, your phone's OEM Android may include a PDF printing service. 
However, having Save to PDF built in to FF is more convenient and likely produces better image quality. It is unavailable now, but slated to be reimplemented at a later date.
That's where we're at, right?

**Comment by headbank** [comment]
@madb1lly Thanks for that, I haven't yet found any authoritative info on the built-in functionality - but it's already clear, sad to say, that OEMs are not all exposing it anyway.

I have seen Google Drive mentioned previously as a favoured solution for PDFing on Android, and that is in my Share menu for Fenix, I imagine it works better than the app I tried but as you say it's only a workaround.

@FunkyBrewster that's a pretty accurate summary (although can't speak as to image quality).

Regarding saving pages I can only see #219 and I'm not even sure whether the scope of that is actual export of  .html (or .war?) files, or simply a built-in offline page cache. You might ask there for clarification on that point.

**Comment by jaqueaux** [comment]
reading through this thread, i did not see any "eta" on correcting this issue. i have always been able to go to "Page > Save as PDF" and a pdf file is then saved to my "Download" folder. is there an "eta" for this fix ?

**Comment by jwatt** [comment]
For reference: now that Firefox Mobile has moved to GeckoView, presumably GeckoView [bug 1659819](https://bugzilla.mozilla.org/show_bug.cgi?id=1659819) needs to be fixed before the Save to PDF option can be implemented.

**Comment by BearOve** [comment]
> So to summarize, you can work around this issue by using the Share function to another app that will output to pdf. If you're lucky, your phone's OEM Android may include a PDF printing service.

This is not correct, at least in the current state. The share option just sends the url to the specified app, and in many (if not most) cases this will not give what you wanted.
* if you are logged in, the link will not work
* if the page has cookies, the GDPR popup will hide the contents behind itself
* if you expand a section or do anything on the page, this will not be reflected
* and so on

This means the only time the share option would be useful is if you sent it to a different browser with print or pdf support

**Comment by cpommier** [comment]
Just another comment to underline the importance of that functionality

**Comment by mohibo** [comment]
> 
> 
> > Desktop browsers let you download their html bundles or let you save an image through the developer tools.
> 
> Firefox 81 for desktop let you save webpages as PDF. 😉

this is not the desktop's github

**Comment by cadeyrn** [comment]
@mohibo: What you quoted was the response to another user's comment who wrote about the Desktop Firefox so I don't know what exactly you want to tell us with your comment to this several months old comment.

**Comment by mohibo** [comment]
> 
> 
> @mohibo: What you quoted was the response to another user's comment who made a comparison to the Desktop Firefox and I don't know why it should be wrong to compare Firefox on different platforms. So I can't see what exactly you want to tell us with your comment to this several months old comment. It was a valid part of the discussion.

we are talking about android,not desktop

**Comment by cadeyrn** [comment]
Everyone understands that. Do you understand the context of these comments? It's not wrong to look what Firefox is able to do on platform A and ask: Why not also on platform B? If Firefox is able to save documents as PDFs on Desktop it can be an argument for bring this feature to Android as well. It doesn't mean that every feature makes sense or should be done on every platform but it's more than okay to have this conversation at least.

**Comment by mohibo** [comment]
> 
> 
> Everyone understands that. But do _you_ understand the context of these comments? It's not wrong at all to look what Firefox is able to do on platform A and ask: Why not also on platform B or what are the alternatives? If Firefox is able to save documents as PDFs on Desktop it _can_ be an argument for bringing this feature to Android as well. It doesn't mean that every Desktop feature makes sense or should be done on Android as well but it's more than okay to have this conversation at least. In contrast to your destructive comments these comments contribute to the discussion about this feature for **Android**.

fennec could do it,fenix can't

**Comment by FunkyBrewster** [comment]
@BearOve Thanks for the clarification. So we're basically SOL on this right now. This new fenix has gotten significantly more stable and has more addons than when it came out, but the lack of print to PDF is still a glaring omission. I really hope it comes back, soon.

**Comment by L31FY** [comment]
> @BearOve Thanks for the clarification. So we're basically SOL on this right now. This new fenix has gotten significantly more stable and has more addons than when it came out, but the lack of print to PDF is still a glaring omission. I really hope it comes back, soon.

I completely agree, and if my trusty PDF viewer app didn't include this feature through the share menu as a workaround for me personally, I don't know what I would do as it's an accessibility issue for me! I often need to zoom in further than allowed or use a reader app which can read from the PDF. This is a continued glaring ommission of a necessary feature and an accessibility function even.

**Comment by singhnsk** [comment]
Has there been any progress in it?
I hate having to use Edge just so I can print the webpage to a PDF. Why isn't fox still incorporating at least the print option so that I can use OS's print to pdf service?

**Comment by madb1lly** [comment]
Hi @singhnsk, 👋
I agree, very frustrating. I send pages to a desktop Firefox and print to PDF from there, which is time consuming and definitely far from convenient, and if the page changes during the time I viewed on mobile and when I can view on desktop then it's too late.
Cheers 🙂

**Comment by gbdomubpkm** [comment]
I'm in the experimentation phase of '''safe Android''' because I don't usually use it, so it's a feedback from several smartphones browsers on this point :

I understood that this printing mechanism in a pdf file no longer exists (I don't know since when exactly, somebody know ?), whereas it existed 'before': it is indeed a great lack and even more than that for a browser of this reputation!!

Fortunately, i noticed this mechanism works well for me under  Brave  and Privacy browser which can be useful in case of need.

**Comment by 2nanfer** [comment]
It is an outrage that the print to pdf function is not back there yet. Issue can only be solved thru addons, if we need addons for every function...you know what I mean

**Comment by Vlad-Sol** [comment]
> It is an outrage that the print to pdf function is not back there yet. Issue can only be solved thru addons, if we need addons for every function...you know what I mean

Which addon do you mean?

**Comment by vxcamiloxv** [comment]
this is the way to lose users and reaffirm chrome monopoly

**Comment by Rodze** [comment]
> Why isn't fox still incorporating at least the print option so that I can use OS's print to pdf service?

Because Mozilla is a dying company unfortunately. It's sad to admit it, but that's the real reason Firefox for Android doesn't have a basic browser feature such as to save webpages.

**Comment by qoheniac** [comment]
Sorry for adding yet another off-topic comment but I just wanted to make sure you developers know that most people following this thread are super thankful that you gift the world with a non-profit browser that really doesn't need to hide behind competitors.

**Comment by mohibo** [comment]
And still not a single update about it

**Comment by qoheniac** [comment]
There are [200](https://bugzilla.mozilla.org/buglist.cgi?o1=equals&v2=new&f2=bug_status&v1=P2&product=GeckoView&o2=equals&f1=priority&query_format=advanced) other bugs of equal priority. Give it some time.

**Comment by FunkyBrewster** [comment]
They've already had more than enough time to reinstate the save to pdf feature. It worked fine for years before it was removed by the big update last year and it's way overdue to comeback. Not being able to save pages is a major usability problem.

**Comment by L31FY** [comment]
> There are [200](https://bugzilla.mozilla.org/buglist.cgi?o1=equals&v2=new&f2=bug_status&v1=P2&product=GeckoView&o2=equals&f1=priority&query_format=advanced) other bugs of equal priority. Give it some time.

Sorry, but I think a lot of those things aren't equal priority. Their priorities are screwed up honestly. They've focused heavily on graphical bugs or easy fixes versus the actually important stuff. Still many add-ons won't "work". You have to make your own collection and use the Nightly build. We are still missing basic usability features as this. Tabs still reload wildly for no reason. This thing receives an update almost daily, sometimes even multiple, and yet the real issues aren't ever touched, so please tell me what they're actually "fixing".

**Comment by mohibo** [comment]
> 
> 
> > Desktop browsers let you download their html bundles or let you save an image through the developer tools.
> 
> Firefox 81 for desktop let you save webpages as PDF. 😉

thanks for the useless information

**Comment by FunkyBrewster** [comment]
It should probably be noted that Chrome for Android now has a working Print to pdf feature. Someone mentioned Brave had it, so it must be baked into Android Chromium now. Reopening the page in Chrome is now my unfortunate workaround.

**Comment by nbmrjuhneibkr** [comment]
Using Chrome to save pages.
Using Chrome to view source.
What other basic web browsing features have been removed from Firefox, but are present in Chrome?

**Comment by benklaasen** [comment]
Can we please stop with the low-quality complaint comment spam. I agree, it's a tragedy what's happening to Firefox. This is all entirely self-inflicted.

What I've done is uninstall this latest Firefox for Android, installed F-Droid, installed Fennec version 68.12.0 from F-Droid and then "Ignore all updates" for Fennec on the Fennec page in F-Droid.

**Comment by nbmrjuhneibkr** [comment]
That's doesn't solve anything. Internet is constantly evolving, so any web browser will eventually become useless without updates. Instead of suggesting to keep using obsolete software indefinitely, we should be pushing for its features to be preserved and expanded in new projects.

**Comment by julienw** [comment]
Hey folks,

Adding "+1" comments or explaining why you think Firefox is terrible doesn't help. Do you think any developer will want to fix a bug that's important to you after you rant at their face? I don't think so.

So, please comment only with useful information, such as if you want to contribute. As said before the first step is to implement [bug 1659819](https://bugzilla.mozilla.org/show_bug.cgi?id=1659819).

I also want to stress that at Mozilla we have [Community Participation Guidelines](https://www.mozilla.org/about/governance/policies/participation/), they exist so that the Mozilla projects stay safe places. This issue doesn't look like a safe place at the moment with all these rants.

It would be a shame to lock the conversation but we may need to do it if this goes on.

Thanks all for your help.

**Comment by L31FY** [comment]
> Hey folks,
> 
> Adding "+1" comments or explaining why you think Firefox is terrible doesn't help. Do you think any developer will want to fix a bug that's important to you after you rant at their face? I don't think so.
> 
> So, please comment only with useful information, such as if you want to contribute. As said before the first step is to implement [bug 1659819](https://bugzilla.mozilla.org/show_bug.cgi?id=1659819).
> 
> I also want to stress that at Mozilla we have [Community Participation Guidelines](https://www.mozilla.org/about/governance/policies/participation/), they exist so that the Mozilla projects stay safe places. This issue doesn't look like a safe place at the moment with all these rants.
> 
> It would be a shame to lock the conversation but we may need to do it if this goes on.
> 
> Thanks all for your help.


Threatening to lock the conversation because honest criticism is taking place doesn't bode well and only compounds what is going on here that users aren't being listened to or taken seriously. This isn't rants. This is feedback. Can we get an actual response as to why basic features like print are still gone and an ETA on a return please? That last sentence is frankly condescending by the way. Nobody has cursed anyone out or said anything unnecessarily inflammatory. They've voiced honest concerns and made true statements.

**Comment by FunkyBrewster** [comment]
> I also want to stress that at Mozilla we have [Community Participation Guidelines](https://www.mozilla.org/about/governance/policies/participation/), they exist so that the Mozilla projects stay safe places. This issue doesn't look like a safe place at the moment with all these rants.

[From the Community Guidelines page:]
>Be Direct but Professional
>
>We are likely to have some discussions about if and when criticism is respectful and when it’s not. We must be able to speak directly when we disagree and when we think we need to improve. We cannot withhold hard truths. Doing so respectfully is hard, doing so when others don’t seem to be listening is harder, and hearing such comments when one is the recipient can be even harder still. We need to be honest and direct, as well as respectful. 

I don't see any comments here that clearly fall under the Unacceptable Behavior section. People are upset because this is an issue that makes FF unusable in situations where it used to be just fine, and on top of that there's a lack of communication about fixing it. Web searches and help/support links are leading people here looking for help.

I understand just complaining about something doesn't help, but for the average user who wants to use FF as their main browser and not Chromium, there isn't much else we can do except try to detail the issue and why we think it needs higher priority, or at least acknowledgment more often than every 6 months.

**Comment by Dunexus** [comment]
It is important to understand the purpose of this github repository. It is first and foremost a work tool for the devs.

Of course, constructive criticism can help, however, on this issue, there has been several comments only repeating that this feature has been lost in the fennec to fenix transition and that mozilla should be ashamed. Repeating this kind of comment doesn't help implementing a feature at all.

Keep in mind that Fenix is brought to you by a small team working hard on it. There are over 2000 open issues, not all of them can be resolved quickly.

Please keep this issue and this repository clean and professional. :heart:

**Comment by gatopeich** [comment]
I know it won't help but this is really a regression, not a "feature request".
You should have spawned a separate product, or a Beta program, if it really is so different and lacks so many features.

**Comment by L31FY** [comment]
So is this just dead now? Or was it added back silently? Am I missing something with this being marked closed?

**Comment by qoheniac** [comment]
Probably because the issue is upstream.

**Comment by FunkyBrewster** [comment]
Save/Print to PDF hasn't been restored as of v. 89.1.1 (May 28 2021 according to my About menu). Maybe someone was closing the duplicates to consolidate them into this thread, which is still open?

**Comment by L31FY** [comment]
I notice something about the fact PC users aren't really affected by this
though as there's generally a native Print to PDF function in the Desktop
OS of your flavor now via the file menu. Android does not have this to my
awareness or it's been hidden from Firefox and cannot utilize that (as with
the standard page print menu). This leaves a huge gap in feature parity
right now on the mobile end and I think may be one reason this is getting
ignored which brings the point it's not so much an upstream issue at all
but an Android version issue.

**Comment by headbank** [comment]
@L31FY printing to PDF is not OS-level on Linux, Firefox implements it natively there (as do other apps including larger desktop environments - this can leave you with more than one PDF "printer" to choose from!).

There actually is code in Android for print to PDF (we had a brief look at this in comments above, see around Aug '20). But it seems that device-makers rarely make this available in their implementations of Android print service. Besides, until Fenix gains the ability to print to *anything* via the Android print service, that's rather moot.

As far as I can tell [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1659819) is what's meant about "upstream". I don't understand any of it but it certainly seems stalled.

**Comment by Nickoriginal** [comment]
Only reason why I still have Chrome enabled is precisely in "Save as PDF" function. Sad as it may seem, there are lost benefits of Fennec, that will never return, despite their heated discussion. Some of them are listed [here](https://github.com/mozilla-mobile/fenix/issues/13302#issuecomment-671493221) (to prevent mentioning), like Tab Queue, Save Page Offline and In Browser Printing.

**Comment by vehosah** [comment]
any actual update about it

**Comment by SeburoSUMO** [comment]
User feedback:

https://support.mozilla.org/en-US/questions/1345712
https://support.mozilla.org/en-US/questions/1345703

**Comment by vehosah** [comment]
And it sent us back to github again

Στις Δευ, 2 Αυγ 2021, 12:58 μ.μ. ο χρήστης Paul Wright <
***@***.***> έγραψε:

**Comment by vehosah** [comment]
> User feedback:
> 
> https://support.mozilla.org/en-US/questions/1345712
> https://support.mozilla.org/en-US/questions/1345703

And the feedback sent us back to github again, without any updates

**Comment by cadeyrn** [comment]
> And the feedback sent us back to github again, without any updates

What kind of update did you expect from these links? 🤔 These are just other feature requests for this feature from Mozilla's support platform support.mozilla.org.

**Comment by vehosah** [comment]
Then why he wrote these links then?

Στις Δευ, 2 Αυγ 2021, 3:53 μ.μ. ο χρήστης Sören Hentzschel <
***@***.***> έγραψε:

**Comment by Nickoriginal** [comment]
@vehosah, user feedback can be (and should be) provided here on Github, to better understand regular users opinion, desire and what needs to be developed for them. Without writing and listening this, no product or service can exist.

**Comment by Nickoriginal** [comment]
Also please reply here directly (not via email) to keep your replies clean.

**Comment by vxcamiloxv** [comment]
Firefox is amazing, thank for all the developer and contributors, I hope have this feature soon to make better FIrefox!

**Comment by MikeCheque** [comment]
[So I found this....](https://searchfox.org/mozilla-central/rev/9fab6c0021ed9b103a70bf5c296b45949a2950bc/browser/components/extensions/parent/ext-tabs.js#1279) 

Can anyone comment on this? Was  <FrameLoader.print().> ever implemented?

**Comment by MikeCheque** [comment]
[This said that](https://searchfox.org/mozilla-central/rev/9fab6c0021ed9b103a70bf5c296b45949a2950bc/browser/components/extensions/parent/ext-tabs.js#1279)  `FrameLoader.print().` and a web-wrapper ought to be used. Any update on this?

**Comment by L31FY** [comment]
> Only reason why I still have _other browser_ installed is precisely in "Save as PDF" function. Sad as it may seem, there are lost benefits of Fennec, that will never be implemented again, despite their heated discussions. Some of them are listed [here](https://github.com/mozilla-mobile/fenix/issues/13302#issuecomment-671493221), like "Tab Queue", "Save Page Offline", "In Browser Printing", "View Page Source" and, of course "Save as PDF".
> I do not blame the developers or the Mozilla itself for this, but I think that no one is ready to implement them yet.

It really seems like they can implement a bunch of things nobody asked for though and that's why it is so frustrating. That can have a bit of responsibility placed for misplaced priorities on their end and not using their time to even attempt it in any way apparently.

**Comment by gatopeich** [comment]
Why not make a plugin based on https://openbase.com/js/pdfkit?

**Comment by SeburoSUMO** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1350541

**Comment by vehosah** [comment]
> > Desktop browsers let you download their html bundles or let you save an image through the developer tools.
> 
> Firefox 81 for desktop let you save webpages as PDF. 😉

And? How is this suppose to help?

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1359330

**Comment by L31FY** [comment]
We continue to get graphical changes nobody requested that I see while this is effectively being censored if It's mentioned along with no real response. Priority should go to function first. This is a function. A rather necessary one at that.

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1360136

**Comment by headbank** [comment]
> User feedback: https://support.mozilla.org/en-US/questions/1360136

@plwt Props for the work you're doing relaying these support posts, but I think this one falls outside this ticket. The feature that poster is asking about, "long screenshots", is quite different from PDF output: it's simply a continuous, non-paginated raster screengrab of the whole page as rendered in the browser (this feature is found in other browsers). Certainly the code-path to implement it would be quite different.

There is a recently added ticket here, #22049 specifically asking for that feature.

**Comment by plwt** [comment]
I have included it under that issue.  We have users looking for this feature that was in both Fennec and Firefox Lite as a means to save content offline.  Either solution would make a lot of people very happy.

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1360316

**Comment by headbank** [comment]
> I have included it under that issue. We have users looking for this feature that was in both Fennec and Firefox Lite as a means to save content offline. Either solution would make a lot of people very happy.

It's no doubt true that many of us would be happy to have just one out of this ticket, #22049, or even #219 for that matter, implemented and have some method of saving a page. However the author of that support thread isn't one of them (see his [response](https://support.mozilla.org/en-US/questions/1360136#answer-1464332)) and now that thread got bogged-down with responses that are off-topic and unhelpful to him (this might happen anyway ofc, but let's not help it along). Ditto on the matching github ticket, and here too (and I'm making it worse, I know lol)

We don't discuss #219 here, and the same should be true for #22049, because these tickets are (supposed to be) defined by the discrete workloads that they present to devs to do (or not do). Where they cross-reference, it should be because the _workloads_ are related to one another.

Conversely, when a new ticket is raised that asks for work already covered in an existing ticket but for novel and distinct reasons, we don't leave the new ticket to stand on that basis; we close it and direct the user to the existing ticket (adding their use-case to that pile).

I respect your opinion on this (and again I HUGELY respect the legwork you're putting in and hope it continues) but I hope you'll give this some consideration. I worry I come off like a squeaky wheel here but my desires for all tickets here are (a) they provide focused and well-argued call-to-arms and plan of action for a dev to read and hopefully choose to take up; (b) users who arrive here by their own efforts or yours can quickly find out where they stand without too much extraneous chaff to wade through; (c) where other tickets' workloads may interplay, all parties gain awareness of that and can spread their input across that base as appropriate. And in this instance I just felt none of those boxes are ticked.

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1362323

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1364060

**Comment by headbank** [comment]
Some movement [on the Bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1659819)...

**Comment by qoheniac** [comment]
In the meantime, this might be helpful to some here: https://github.com/billthefarmer/print

**Comment by headbank** [comment]
@qoheniac I don't think this helps as it appears to simply print a file passed to it. Fenix has no option to do this: it can only share the page's URL.

**Comment by headbank** [comment]
@qoheniac Thanks for clarifying. The app's README says you can "share _a file or text_ in another app" - no mention that it can also take a URL and download by itself the linked resource.

If it can do that then it helps for static, consistently-rendering pages, yes. However the pages we really need to print/record are often ones that are behind a login (think receipts/confirmations etc), or otherwise rendered uniquely based on the current session within Fenix. These cannot be replicated by requesting the same URL from within an external app; often the response will be different or nonexistent.

All the same this will be quite helpful for some while we await proper resolution (which may finally be in the works now), so thanks for mentioning it.

**Comment by vehosah** [comment]
> Some movement [on the Bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1659819)...

At last

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1369883

**Comment by gungoksu** [comment]
Printing to pdf is a key function of a browser when doing research or just for use as reference later. This regression is a major issue for me. Looking at this thread, though still marked as new on top of page, it is from 2019. I find it unbelievable anyone thought removing printing was a good idea. Looking at all the other cosmetic stuff being worked on, I do not believe printing will be resolved any time "soon" if ever. I am typing this from chrome as I have now uninstalled Firefox and will look at other browsers. What a shame. Oh and I'll take a print of this page for records for if it disappears.

**Comment by L31FY** [comment]
> Printing to pdf is a key function of a browser when doing research or just for use as reference later. This regression is a major issue for me. Looking at this thread, though still marked as new on top of page, it is from 2019. I find it unbelievable anyone thought removing printing was a good idea. Looking at all the other cosmetic stuff being worked on, I do not believe printing will be resolved any time "soon" if ever. I am typing this from chrome as I have now uninstalled Firefox and will look at other browsers. What a shame. Oh and I'll take a print of this page for records for if it disappears. 

Look out or you'll get flagged for hate since you don't agree with their decisions. See above a bit how they tried to silence the negative feedback with that nonsense. I've tried several workarounds now that all have various failings and it's encouraging to finally maybe see some progress natively on putting it back, but priorities of the team are definitely wrong and I don't care who wants to disagree with that. It was a mistake to remove it to begin with. Full stop. It's also been a mistake to not focus on features and usability and instead repeatedly rework cosmetic aspects that were neither broken or asked to be done. It's clear that the severity or popularity of a feature or bug like this means basically nothing in the grand pipe of things to be done and is only a placeholder. Unless it's actually unusable and crashed completely like some of the Nightly builds have, they won't bother to care enough unless it suits what they want to do despite claiming to listen to the community. We are ignored and have been for some time as far as any decisions go.

**Comment by headbank** [comment]
@gungoksu and @L31FY may be pleased to learn that there is now a target of Fenix 100/101 for this work, see

https://bugzilla.mozilla.org/show_bug.cgi?id=1659819

Note that general printing support is waiting on this (as I understand it, it's the main thing standing in the way)

https://bugzilla.mozilla.org/show_bug.cgi?id=1659818

**Comment by ccscoggins** [comment]
For anyone else looking out for this: According to the bugzilla, this feature went from 'targeted for Fenix 100/101' to 'no need to implement for Fenix 102 [nightly]'.  So uh, we'll still be waiting a couple more updates probably.
The 'no need' phrasing strikes me as kind of odd, because it would seem to imply that this is only an issue that others are dependent on rather than itself a direct (and seemingly fairly often requested) feature, but I might just be reading too much into it.

**Comment by cpeterso** [comment]
> For anyone else looking out for this: According to the bugzilla, this feature went from 'targeted for Fenix 100/101' to 'no need to implement for Fenix 102 [nightly]'. So uh, we'll still be waiting a couple more updates probably. The 'no need' phrasing strikes me as kind of odd, because it would seem to imply that this is only an issue that others are dependent on rather than itself a direct (and seemingly fairly often requested) feature, but I might just be reading too much into it.

You're correct. Bug https://bugzilla.mozilla.org/show_bug.cgi?id=1659819 is adding a "Save as PDF" API in the Gecko engine. The Fenix team will then need to add UI to call the API. That UI work will be tracked in this GitHub issue and is on the Fenix team's 2022 roadmap.

**Comment by headbank** [comment]
The Gecko bug has just been closed (fixed) and if I follow correctly it should be in v103. So the next thing to look out for is, I assume, somebody getting assigned to this bug here. No knowing when that will happen, I think the roadmap mentioned above is not a public document; @cpeterso if I'm wrong, please can you direct us to it?

**Comment by aemjay** [comment]
I can't believe this is still pending three years down the road. Made me switch to Brave and I am still waiting for this to make the change back.

**Comment by cadeyrn** [comment]
@aemjay: Just read the previus comments. They are working on it, making your comment unnecessary bashing.

**Comment by aaronwi** [comment]
what specifically is taking so long to add back code that was removed?

**Comment by cadeyrn** [comment]
I can't answer any detail questions, I am not involved. But they didn't remove anything in this regard. Fenix has never had this feature.

**Comment by petertirrell** [comment]
To be fair Firefox for Android originally *did* have printing and saving to pdf capabilities and then lost them. This should have been treated as a regression, not a feature request.

**Comment by cadeyrn** [comment]
_To be fair_, Fenix is not Fennec, it's a different browser. Therefore it's not a regression but a feature request. But to be honest, it doesn't matter at all how you call it, since Mozilla is working on it.

**Comment by madb1lly** [comment]
_To be fair_ this is the umpteenth time you've lectured us about Fenix being a different browser, the dev team should be left to work on it, yaddayadda. From a user perspective it's _not_ a different browser because Mozilla chose to use _the same_ app store entry for it as they had been using for Fennec, "org.mozilla.firefox", therefore the loss of the Save to PDF, printing and the loss of many other features are indeed regressions and _not_ feature requests. Honestly, were probably very close to breaking app store rules when they replaced Fennec with Fenix.

I'm glad to hear that Mozilla are now working on this :)

**Comment by aaronwi** [comment]
> _To be fair_, Fenix is not Fennec, it's a different browser. Therefore it's not a regression but a feature request. But to be honest, it doesn't really matter how we call it, since Mozilla is working on it. At this point it won't make a difference.

No one cares about the codenames or backend, you have to look at the users perspective, which clearly no one does both on android or on desktop versions, firefox is firefox, and it's terrible now. Look at brave, a completely complete user interface on android, windows and linux, everything works, printing works, even support for 'web3, crypto stuff', what has firefox done in the last decade? broken webextensions?. I just switched, and am not going back.

**Comment by L31FY** [comment]
> To be fair Firefox for Android originally *did* have printing and saving to pdf capabilities and then lost them. This should have been treated as a regression, not a feature request.

We were treated as hostile combatants for suggesting this before and it seems to be so again. Be careful or we will be banned for speaking that we don't love something about this new direction. This team cannot take criticism even of the most benign kind without lashing out and accusing you of flaming them or being hateful when you just report a bug or want something that broke fixed. It's also valid to question timeline on things and wonder why seemingly simple issues or feature regressions are taking extremely long to fix but they consider that to be hateful and not a point of discussion as can be seen above so I've gone to keeping my mouth shut as censoring was their intentions from the clear ban threats.

**Comment by aaronwi** [comment]
> > To be fair Firefox for Android originally _did_ have printing and saving to pdf capabilities and then lost them. This should have been treated as a regression, not a feature request.
> 
> We were treated as hostile combatants for suggesting this before and it seems to be so again. Be careful or we will be banned for speaking that we don't love something about this new direction. This team cannot take criticism even of the most benign kind without lashing out and accusing you of flaming them or being hateful when you just report a bug or want something that broke fixed. It's also valid to question timeline on things and wonder why seemingly simple issues or feature regressions are taking extremely long to fix but they consider that to be hateful and not a point of discussion as can be seen above so I've gone to keeping my mouth shut as censoring was their intentions from the clear ban threats.

It appears they are actively trying to destroy the firefox brand to push everyone over to chrome, maybe google is paying the devs to do so.

**Comment by Nerixyz** [comment]
Could you please discuss this somewhere else?

Everyone subscribed to this issue/interested in this issue gets a GitHub notification for every (offtopic) comment. 

**If you want to show interest in having this feature added, react with 👍 on [the original issue comment](https://github.com/mozilla-mobile/fenix/issues/3709#issue-461603457)** and most importantly, **don't comment +1**. If you want to get updates on this issue, click the _Subscribe_ button.

**Comment by 424344** [comment]
@Nerixyz  I don't agree, all comments here are relevant, even those ranting about the apathy of mozilla developers: they must know that users are not happy at all. The only off topic comments here are yours (and mine).

**Comment by cadeyrn** [comment]
>  all comments here are relevant, even those ranting about the apathy of mozilla developers: they must know that users are not happy at all. The only off topic comments here are yours (and mine).

It seems that you didn't unterstand that they're already working on the feature. Even if it were really necessary to write such comments (which I think is a pretext to vent frustration) it doesn't accelerate anything. They won't develop faster if they get offended. It only causes notification spam for all the people who subscribed to this issue because they're interestered in getting status updates. And by the way, the time the developers have to waste for such comments is time they can't invest in developing. So it's in the **best interest of everyone who is really interested in the feature** to let the developers do their work and refrain from further bashing.

Some people should really learn that the only way to achieve something is to behave in a friendly and respectful way. Anti-social behavior and offending the developers were never a helpful behaviour.

**Comment by FunkyBrewster** [comment]
> It seems that you didn't unterstand that they're already working on the feature. Even if it were really necessary to write such comments (which I think is a pretext to vent frustration) - at this point it can't accelerate anything.

This thread was started in June, 2019, which means they've been "working" on it for over three years. Three years with zero progress or updates. People aren't just venting, they are users who want to support FF but are expressing that time has run out. I don't want a chromium-only web, but I'm at the point of biting the bullet and switching to what works.

**Comment by jonalmeida** [comment]
As noted in https://github.com/mozilla-mobile/android-components/pull/12277#pullrequestreview-1072737598, due to some technical limitations we can integrate this feature in Fenix and keep it behind a nightly flag until we can get a fix to improve this situation.

**Comment by IGS-GIT** [comment]
Was tryna save a recipe, not find an internet battle about a missing feature. Gawd damn...

**Comment by julienw** [comment]
It's now available at least on the Nightly version of Firefox. You can find it at the bottom of the share menu.

**Comment by IGS-GIT** [comment]
> It's now available at least on the Nightly version of Firefox. You can find it at the bottom of the share menu.

How about when I press a print option on a page? Will it know to bring up the print screen?

**Comment by L31FY** [comment]
> > It's now available at least on the Nightly version of Firefox. You can find it at the bottom of the share menu.
> 
> How about when I press a print option on a page? Will it know to bring up the print screen?

That's still not functional and the print menu item is still missing. Clicking a print button often does nothing or loads a more print friendly page only in this case. 

I still don't understand why such oversight was allowed to happen and then be doubled down on, even calling the critical users the ones in the wrong, about basic usability and accessibility features. Insulting the userbase with disrespect and hostility and removing features is a direction I hope the team moves far away from.

**Comment by kbrosnan** [comment]
bz-1659818 is for window.print.

**Comment by smaudet** [comment]
@L31FY Firefox has had a dramatic loss of userbase - some of it was OFC domination by Google Chrome, but I wonder if the developers have become...worse.
